File: frontend/src/types/driver.ts
export interface DriverCatalogEntry {
  key: string
  name: string
  vendor: string
  direction: 'Southbound' | 'Northbound'
  config_schema: Record<string, unknown>
  tag_schema: Record<string, unknown>
}

export interface DriverInstance {
  id: string
  runtime_node_id: string
  pea_id: string
  driver_key: string
  display_name: string
  state: 'Created' | 'Configured' | 'Running' | 'Stopped' | 'Error'
  config: Record<string, unknown>
  tag_groups: TagGroup[]
  last_error?: string | null
  created_at: string
  updated_at: string
}

export interface DriverSchemaPayload extends DriverCatalogEntry {
  source?: 'neuron' | 'builtin'
}

export interface DriverOperationRecord {
  tag_id: string
  tag_name: string
  value: unknown
  ok: boolean
  message?: string | null
  timestamp: string
}

export interface DriverStatusSnapshot {
  driver_id: string
  node_name: string
  state: DriverInstance['state']
  remote_running?: boolean | null
  remote_link?: number | null
  remote_rtt?: number | null
  last_error?: string | null
  last_read?: DriverOperationRecord | null
  last_write?: DriverOperationRecord | null
  updated_at: string
}

export interface TagGroup {
  id: string
  name: string
  description?: string | null
  tags: DriverTag[]
}

export interface DriverTag {
  id: string
  name: string
  address: string
  data_type: 'Bool' | 'Int16' | 'Uint16' | 'Int32' | 'Uint32' | 'Float32' | 'Float64' | 'String'
  access: 'Read' | 'Write' | 'ReadWrite'
  scan_ms?: number | null
  attributes: Record<string, unknown>
}

File: backend/shared/src/domain/driver.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverCatalogEntry {
    pub key: String,
    pub name: String,
    pub vendor: String,
    pub direction: DriverDirection,
    pub config_schema: serde_json::Value,
    pub tag_schema: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DriverDirection {
    Southbound,
    Northbound,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverInstance {
    pub id: String,
    pub runtime_node_id: String,
    pub pea_id: String,
    pub driver_key: String,
    pub display_name: String,
    pub state: DriverInstanceState,
    pub config: serde_json::Value,
    pub tag_groups: Vec<TagGroup>,
    pub last_error: Option<String>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverStatusSnapshot {
    pub driver_id: String,
    pub node_name: String,
    pub state: DriverInstanceState,
    pub remote_running: Option<bool>,
    pub remote_link: Option<i64>,
    pub remote_rtt: Option<i64>,
    pub last_error: Option<String>,
    pub last_read: Option<DriverOperationRecord>,
    pub last_write: Option<DriverOperationRecord>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverOperationRecord {
    pub tag_id: String,
    pub tag_name: String,
    pub value: serde_json::Value,
    pub ok: bool,
    pub message: Option<String>,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DriverInstanceState {
    Created,
    Configured,
    Running,
    Stopped,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TagGroup {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub tags: Vec<DriverTag>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverTag {
    pub id: String,
    pub name: String,
    pub address: String,
    pub data_type: DriverDataType,
    pub access: TagAccess,
    pub scan_ms: Option<u32>,
    pub attributes: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DriverDataType {
    Bool,
    Int16,
    Uint16,
    Int32,
    Uint32,
    Float32,
    Float64,
    String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum TagAccess {
    Read,
    Write,
    ReadWrite,
}
